### PR TITLE
Assignment5 @ lsobtafo

### PR DESCRIPTION
**Pull request by slanante**

### CODE REVIEW

**Comment by BenagainB** [comment]
Create a class library project called "GenericsHomework.". ✔
Create a node class that can contain a value of any type and points to the next node and traversing the next node points back to the first item.✔
Define the Node class✔
Include a constructor that takes a value. (No validation is necessary on the value). ✔
Add a ToString() override that writes out the value's ToString() result. ✔
Add a Next property that references the next node or else refers back to itself if there are no other nodes in the list. ✔
The Next property should be non-nullable (careful to follow the non-nullable property guidelines) ✔
The Next property setter should be private. ✔
Add an Append method that takes a value and appends a new Node instance after the current node (by invoking the Next property). ✔
Add a Clear method that effectively removes all items from a list except the current node. Pay attention as to whether you should be concerned with the following:
Whether it is sufficient to only set Next to itself ✔
Whether to set the removed items to circle back on themselves. In other words, whether to close the loop of the removed items. (Provide a test to show why this is required if it is required). ✔
Given there is a circular list of items, provide a comment to indicate whether you need to worry about garbage collection because all the items point to each other and therefore may never be garbage collected. ✔
Create an Exists method to test to see if a value exists in the list. ✔
Throw an error on an attempt to Append a duplicate value. (Make sure you test for this case) ✔
You should not rely on any BCL generic classes for your implementation. ✔
Extra Credit
Do one of the following two options (or both if you want extra, extra credit) :)

Implement a VennDiagram structure that contains n Circles that only contains homogenous reference types of any type. ❌✔
Each circle contains n items and each item can belong to one or more Circle instances.
You are not required to use a Node from earlier in the homework for your Venn diagram implementation.
You are welcome to use existing BCL generic classes for the extra credit.
Implement System.Collections.Generic.ICollection<T> on the Node class ❌✔
Fundamentals
Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

So, I'm still getting used to using generics, but you've used them very cleanly here. Especially like how you wrote your Dispose method inside Node.

File: GenericsHomework/Node.cs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    {
+        if (Content != null)
+        {
+            return Content.ToString()!;
+        }
+        else
+        {
+            return null;
+        }
+
```
Comment (mmwoodfo) [inline GenericsHomework/Node.cs:25]:
```suggestion
return Content.ToString();
```

File: GenericsHomework/Node.cs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    {
+        if (Exists(content))
+        {
+            throw new Exception("The value within Content has already been appended previously.");
+        }
+
+        Node<T> newNode = new Node<T>(content);
+        Node<T> current = this;
+
+        if (NextNode.Equals(this))
```
Comment (mmwoodfo) [inline GenericsHomework/Node.cs:38]: You don't need to split this into an if/else statement, you should be able to just do something such as this:
```
           newNode.Next = Next;
            Next = newNode;
```

File: GenericsHomework/Node.cs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        List<Node<T>> nodes = new List<Node<T>>();
+
+        while (currNode.NextNode != currNode)
+        {
+            currNode = currNode.NextNode;
+            nodes.Add(currNode);
+        }
+        foreach (Node<T> node in nodes)
+        {
+            Dispose(node);
```
Comment (mmwoodfo) [inline GenericsHomework/Node.cs:83]: You do not need to call dispose here

File: GenericsHomework/Node.cs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        }
+        NextNode = this;
+        PrevNode = null;
+    }
+
+    public bool Exists(T content)
+    {
+        Node<T> currNode = this;
+        while (currNode != null)
```
Comment (mmwoodfo) [inline GenericsHomework/Node.cs:92]: When will current node equal null?

File: GenericsHomework/Node.cs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+    Or at least try to...
+
+    */
+    public void Clear()
+    {
+        Node<T> currNode = this;
+        List<Node<T>> nodes = new List<Node<T>>();
+
+        while (currNode.NextNode != currNode)
```
Comment (mmwoodfo) [inline GenericsHomework/Node.cs:76]: You want to stop your loop when you reach the end of your list, you'll reach the end of your list when your `Next` property references back to it's self.

File: GenericsHomeWork.Tests/NodeTests.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
+namespace GenericsHomework.Tests;
+
+[TestClass]
+public class NodeTests
+{
+    [TestMethod]
+    public void Node_Append_Success()
```
Comment (mmwoodfo) [inline GenericsHomeWork.Tests/NodeTests.cs:7]: You should have a test for to string in here as well as some more tests for testing clearing, appending and traversing on different list lengths.

**Pull request closed without merging**